### ISSUE TITLE
Fuzzer: binaryen interpreter and d8 comparison error

### ISSUE THREAD

**Issue by MaxGraey** [WebAssembly/binaryen#3149]
It's not relate to #3138. It seems now it relate to `memory.fill` or other memory operation.

Seed: `15650187477080971485`.

VMs diff:
```diff
 [fuzz-exec] calling func_24
 [fuzz-exec] note result: func_24 => -65
 [fuzz-exec] calling func_26_invoker
-[LoggingExternalInterface logging 1169807931]
-[LoggingExternalInterface logging 1169807931]
+[LoggingExternalInterface logging 1169807907]
+[LoggingExternalInterface logging 1169807907]
 [fuzz-exec] calling func_28
```

Reduced sample:
```wat
(module
  (type $t0 (func))
  (type $t1 (func (param i32)))
  (import "fuzzing-support" "log-i32" (func $fuzzing-support.log-i32 (type $t1)))
  (func $func_19_invoker (export "func_19_invoker") (type $t0)
    (memory.fill
      (i32.const 15)
      (i32.const 14)
      (i32.const -2)))
  (func $func_29_invoker (export "func_29_invoker") (type $t0)
    (call $fuzzing-support.log-i32
      (i32.load8_u
        (i32.const 15))))
  (memory $M0 1 1))
```

**UPDATE**
Last working commit: 1a928bc3ff4b511e81b3f93db8aea872e88abaaf
So it seems this introduced in #3130

**Comment by dcodeIO** [comment]
Hit a similar one: **11266044653066427376**

```wat
wasm-opt a.wasm --all-features --disable-nontrapping-float-to-int --disable-simd --disable-sign-ext --disable-exception-handling --disable-reference-types --disable-exception-handling --disable-gc --disable-multivalue --disable-memory64 --fuzz-exec-before

d8 a.wasm.js --wasm-staging --experimental-wasm-eh --experimental-wasm-simd --experimental-wasm-reftypes --experimental-wasm-compilation-hints --experimental-wasm-return-call -- a.wasm
```

```wat
(module
 (type $none_=>_none (func))
 (type $i32_=>_none (func (param i32)))
 (type $none_=>_i32 (func (result i32)))
 (type $i32_i32_f64_=>_i32 (func (param i32 i32 f64) (result i32)))
 (import "fuzzing-support" "log-i32" (func $fimport$0 (param i32)))
 (memory $0 (shared 1 1))
 (export "func_23" (func $2))
 (export "func_44" (func $0))
 (export "func_45_invoker" (func $1))
 (func $0 (result i32)
  (local $0 i32)
  (memory.fill
   (i32.const 1)
   (local.tee $0
    (i32.load8_s
     (i32.const 23)
    )
   )
   (local.get $0)
  )
  (i32.const 0)
 )
 (func $1
  (call $fimport$0
   (i32.load8_u
    (i32.const 14)
   )
  )
 )
 (func $2 (param $0 i32) (param $1 i32) (param $2 f64) (result i32)
  (drop
   (i32.atomic.rmw16.or_u offset=22
    (i32.const 0)
    (i32.const -86)
   )
  )
  (i32.const 0)
 )
)
```

That was on my `ref.eq` branch, though, so not sure if the seed will work on master.

**Comment by kripken** [comment]
I agree this does look like a regression from Memory64 #3130 as @MaxGraey said, cc @aardappel 

Specifically in the interpreter, what happens in the first testcase here is that two `int64` values are added (`15` and `18446744073709551614`) and they overflow, so we don't detect that we should trap. I think the problem is that `getInteger` does a sign extension here.

**Comment by aardappel** [comment]
I don't follow.. `getInteger` indeed shouldn't sign extend (let's add a `getUnsigned` or whatever?) but in this case it only does that for `destVal` which is `15` so it shouldn't matter? the `-2` is loaded over `geti32` and a cast to `uint32_t` before it becomes a `uint64_t`, so should be `0xFFFFFFFE` instead?

Where would be a good test add for this? Spec tests surprisingly don't seem to have tests with a negative size?

https://github.com/WebAssembly/binaryen/blob/ee00f647750f23e6c24e67424bafab39b244c835/src/wasm-interpreter.h#L2346-L2388

**Comment by kripken** [comment]
Oh, yeah, I might have missed something, I just took a quick look. Anyhow, yeah, it might be the sign/unsign issue you mention.

A new spec test could make sense. If for some reason that isn't practical, adding the reduced testcase in `test/passes/fuzz-exec_O.wast` is an option.

**Comment by aardappel** [comment]
Ok, trying to create a repro case that fails, but adding a test like

```
(assert_trap (invoke "init" (i32.const 0) (i32.const 3) (i32.const -2)))
```

To the existing spec tests just traps as expected.

I don't think I fully understand what is happening in the first example above. It's outputting `1169807907` but should output `1169807931` ? How can that be the result of a `load8_u` instruction in the code? It doesn't even specify a `data` section, so if the `memory.fill` traps as it should, shouldn't all memory be 0, and thus the result of the load also?

I am clearly missing a lot of context here. Anyone care to point me at an example of `memory.fill` that would misbehave?

**Comment by kripken** [comment]
@aardappel You may need to add a declaration of the memory to the first testcase here - not sure why it's omitted (looks like wabt output, maybe it doesn't print it in some cases?).

The output before the first testcase, and the first testcase, are not in sync. The output is from the original thing the fuzzer found, the first testcase is what the reducer found. While it reduced it it changed the output. But the output still diverges between v8 and the interpreter.

To see the failure you can just run the first testcase with `--fuzz-exec`. The first export called should trap before writing anything, so the second will print out something that is not zero, because we did end up writing, and that output is wrong.

**Comment by MaxGraey** [comment]
Fixed in #3176
